### PR TITLE
Web: Redirect to login upon missing session cookie

### LINKED ISSUE #20996: Access Denied - missing session cookie
Expected behavior:
When I open the app such as `testapp.teleport.example.com` and I'm not logged in, I expect to be redirected to the login page and then to the app, all automatically -- and it does happen on the first time.

Current behavior:
Launching the app's URL for the second time after the browser was closed on the phones leads to this error being shown for some reason, instead of the login page:
image

This is rather disruptive for the end-users who just need an access to one of the apps, and be presented with login if needed, instead of having them bother about finding the dashboard and launching the app from there...

Worth noting that this error doesn't seem to happen *every* time, especially the chances of encountering it are lower if a few minutes have already passed.

Bug details:
- Teleport version: Teleport v11.3.1 git:api/v11.3.1-0-gccd6a1b02a go1.19.5
- Recreation steps:
1. Have some app and some user in cluster
2. Login to the app on your phone, tested in Firefox for Android and Safari on iOS
3. Teleport does the login flow as expected
4. Add the app (`https://testapp.teleport.example.com/`) to home screen
5. Close the browser and either open the app from home screen shortcut, or even just reopen the closed browser to its last tab
6. You'd expect either the session to be still logged in, or get redirected to login
7. Instead you get aforementioned error, the page you were redirected to is: `https://teleport.example.com/web/launch/testapp.teleport.example.com?path=%2F`
8. In around 30 seconds or so of watching the error page, you get redirected to actual login apparently, but logging in redirects you to main Teleport dashboard instead of the app......
- Debug logs

### PR DESCRIPTION
**Pull request by kimlisa**
fixes https://github.com/gravitational/teleport/issues/20996

Why: We use [non-persistent (session) cookie](https://cheatsheetseries.owasp.org/cheatsheets/Session_Management_Cheat_Sheet.html#expire-and-max-age-attributes). When you close a browser, session cookies get cleared from the browser. Then when you re-open browser, load the teleport app, try to make a teleport api call to a protected endpoint [WithAuth](https://github.com/gravitational/teleport/blob/3c68797ea3ff5ec3cabc8f4bf9bdd68e641bb36b/lib/web/apiserver.go#L4038), it fails to authn because a cookie wasn't sent with the request.

There is three kind of dialogue shown depending on which webapp route you take.

#### AppLauncher component catches error (from trying to `getFqdn`)
Reproduce (chrome, but works on any browser): 
1) login to teleport and successfully launch an app
2) close the browser
3) reopen browser and launch the app (copy and paste app URL)

image

#### Enterprise: the WaitingRoom catches error (from trying to `getUserContext`)
1) login to teleport
2) close the browser
3) reopen browser and go to some teleport URL

image

#### Open source: the root CatchError catches error (from trying to `getUserPreferences`)
image

Our component tree looks like this:

```ts
<CatchError> // our root error catcher if the childrens did not  handle it
  <Authenticated>  // <----- where we should be checking validity of cookie and session
    <AppLauncher/>
    <WaitingRoom>   // only in enterprise
        {... childrens subjected to waiting room}
    </WaitingRoom>
  </Authenticated>
</CatchError>
```

Prior to this PR, in the `Authenticated` layer we were only checking for valid session stored in local storage. But gives false positives. Now we just make the api call to the `/status` endpoint right away and fail earlier in the tree (instead of 15 seconds later from starting an interval timer that polls this endpoint)

All the above scenarios, the user will be redirected to the login right away. On rare occassions where the `/status` just fails to fetch, this is what the user will see:

image

### CODE REVIEW

File: web/packages/teleport/src/components/Authenticated/ErrorDialogue.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import Dialog, {
+  DialogHeader,
+  DialogTitle,
+  DialogContent,
+  DialogFooter,
+} from 'design/Dialog';
+
+import history from 'teleport/services/history';
+
+export function ErrorDialogue({ errMsg }: { errMsg: string }) {
```
Comment (gzdunek) [inline web/packages/teleport/src/components/Authenticated/ErrorDialogue.tsx:28]: `ErrorDialog`?

File: web/packages/teleport/src/components/Authenticated/Authenticated.tsx [hunk 2]
```diff
@@ -39,11 +44,37 @@ const events = [
...
 const Authenticated: React.FC = ({ children }) => {
-  React.useEffect(() => {
-    if (!session.isValid()) {
-      logger.warn('invalid session');
-      session.clear();
-      history.goToLogin(true);
+  const { attempt, setAttempt } = useAttempt('processing');
+
+  useEffect(() => {
+    const checkUserIsAuthenticated = async () => {
```
Comment (gzdunek) [inline web/packages/teleport/src/components/Authenticated/Authenticated.tsx:50]:
```suggestion
    const checkIfUserIsAuthenticated = async () => {
```

File: web/packages/teleport/src/components/Authenticated/Authenticated.tsx [hunk 3]
```diff
@@ -39,11 +44,37 @@ const events = [
...
+      try {
+        if (!session.isValid()) {
+          logger.warn('invalid session');
+          session.clear();
+          history.goToLogin(true);
+          return;
+        }
```
Comment (gzdunek) [inline web/packages/teleport/src/components/Authenticated/Authenticated.tsx:57]: We can move it above `try`.

File: web/packages/teleport/src/components/Authenticated/Authenticated.tsx [hunk 4]
```diff
@@ -55,13 +86,17 @@ const Authenticated: React.FC = ({ children }) => {
...
+  }
 
-  if (!session.isValid()) {
-    return null;
+  if (attempt.status === 'failed') {
+    return <ErrorDialogue errMsg={attempt.statusText} />;
   }
 
-  return <>{children}</>;
+  return null;
```
Comment (gzdunek) [inline web/packages/teleport/src/components/Authenticated/Authenticated.tsx:99]: WDYT of showing an `<Indicator />` during loading?

File: web/packages/teleport/src/services/websession/websession.ts [hunk 5]
```diff
@@ -190,14 +195,23 @@ const session = {
...
       if (err.response.status == 403) {
         this.logout();
       }
     });
   },
 
+  /**
+   * validateCookieAndSessionFromBackend makes an authenticated request
+   * which just checks if the cookie sent from browser is valid and user
+   * session is still valid.
```
Comment (gzdunek) [inline web/packages/teleport/src/services/websession/websession.ts:209]: I'd shorten it a bit:
```suggestion
   * which checks if the cookie and the user session are still valid.
```

**Review by gzdunek — APPROVED** [review]
LGTM, I left a few minor comments.

**Review by ryanclark — APPROVED** [review]

**Pull request merged**